src/pending_connection/connect.rs

```diff
@@ -1,208 +1,261 @@
-    SocketType, SrtControlPacket, SrtHandshake, SrtShakeFlags,
-};
+use log::warn;
+
+use crate::packet::*;
 use crate::util::get_packet;
-use crate::{Connection, ConnectionSettings, SocketID, SrtVersion};
+use crate::{Connection, ConnectionSettings, DataPacket, SeqNumber, SocketID, SrtVersion};
 
-pub async fn connect<T>(
-    sock: &mut T,
+use ConnectError::*;
+use ConnectState::*;
+
+pub struct ConnectConfiguration {
     remote: SocketAddr,
     local_sockid: SocketID,
     local_addr: IpAddr,
     tsbpd_latency: Duration,
-    _crypto: Option<(u8, String)>,
-) -> Result<Connection, Error>
-where
-    T: Stream<Item = Result<(Packet, SocketAddr), Error>>
-        + Sink<(Packet, SocketAddr), Error = Error>
-        + Unpin,
-{
-    info!("Got hanshake from {}", remote);
-    let mut send_interval = interval(Duration::from_millis(100));
-
-    let (timestamp, hs_info) =
-        get_initial_handshake(sock, local_addr, local_sockid, remote).await?;
-
-    // send back a packet with the same syn cookie
-    let pack = Packet::Control(ControlPacket {
-        dest_sockid: SocketID(0),
-        timestamp,
-        control_type: ControlTypes::Handshake(HandshakeControlInfo {
-            shake_type: ShakeType::Conclusion,
-            socket_id: local_sockid,
-            info: HandshakeVSInfo::V5 {
-                crypto_size: 0, // TODO: implement
-                ext_hs: Some(SrtControlPacket::HandshakeRequest(SrtHandshake {
-                    version: SrtVersion::CURRENT,
-                    // TODO: this is hyper bad, don't blindly set send flag
-                    // if you don't pass TSBPDRCV, it doens't set the latency correctly for some reason. Requires more research
-                    peer_latency: Duration::from_secs(0), // TODO: research
-                    flags: SrtShakeFlags::TSBPDSND | SrtShakeFlags::TSBPDRCV, // TODO: the reference implementation sets a lot more of these, research
-                    latency: tsbpd_latency,
-                })),
-                ext_km: None,
-                // ext_km: self.crypto.as_mut().map(|manager| {
-                //     SrtControlPacket::KeyManagerRequest(SrtKeyMessage {
-                //         pt: 2,       // TODO: what is this
-                //         sign: 8_233, // TODO: again
-                //         keki: 0,
-                //         cipher: CipherType::CTR,
-                //         auth: 0,
-                //         se: 2,
-                //         salt: Vec::from(manager.salt()),
-                //         even_key: Some(manager.wrap_key().unwrap()),
-                //         odd_key: None,
-                //         wrap_data: [0; 8],
-                //     })
-                // }),
-                ext_config: None,
-            },
-            ..hs_info
-        }),
-    });
+}
 
-    sock.send((pack.clone(), remote)).await?;
+#[derive(Clone)]
+pub enum ConnectState {
```

Comment:
If we end up merging this I would like these tuple variants to have some docs, it's not immediately obvious what all of these are.